File: src/main/scala/fppattern/chapter02/sc/factorymethod/CurrencyConverter.scala
package fppattern.chapter02.sc.factorymethod

/**
  * Created by zyong on 2016/11/23.
  */
trait Currency {
  def getConversionRateToIndianRupee: String
}

object CurrencyConverter extends App {

  private object EuroToRupee extends Currency {
    override def getConversionRateToIndianRupee = "82"
  }

  private object DollarToRupee extends Currency {
    override def getConversionRateToIndianRupee = "60"
  }

  private object NoIdea extends Currency {
    override def getConversionRateToIndianRupee = "No Idea"
  }

  // the currency factory method
  // Note: Scala if statement is an expression.
  def apply(s: String): Currency = {
    if (s == "Dollar") // same as s.equals("Dollar") in Java
      DollarToRupee
    else if (s == "Euro")
      EuroToRupee
    else
      NoIdea
  }

  val c = CurrencyConverter ("Dollar") // apply method in action
  println(c.getConversionRateToIndianRupee)
}

File: src/main/scala/fppattern/chapter02/ja/factorymethod/CurrencyConverter.java
package fppattern.chapter02.ja.factorymethod;

/**
 * Created by zyong on 2016/11/23.
 */
public class CurrencyConverter {
    private static final String YEN = "Yen";
    private static final String EURO = "Euro";
    private static final String DOLLAR = "Dollar";

    private static final class EuroToRupee implements Currency {
        @Override
        public String getConversionRateToIndianRupee() {
            return "82";
        }
    }

    private static final class DollarToRupee implements Currency {
        @Override
        public String getConversionRateToIndianRupee() {
            return "60";
        }
    }

    private static Currency createCurrencyFor(final String currencyStr) {
        if (currencyStr.equals(DOLLAR)) {
            return new DollarToRupee();
        }
        if (currencyStr.equals(EURO)) {
            return new EuroToRupee();
        }
        throw new IllegalArgumentException("Oops! no idea about <" + currencyStr + ">");
    }

    public static void main(String[] args) {
        System.out.println(createCurrencyFor(DOLLAR).getConversionRateToIndianRupee());
        System.out.println(createCurrencyFor(EURO).getConversionRateToIndianRupee());
        System.out.println(createCurrencyFor(YEN).getConversionRateToIndianRupee());
    }
}